stdlib/public/core/Integers.swift

```diff
@@ -2328,23 +2328,25 @@ extension FixedWidthInteger {
   // that it needs to generate efficient code.
   @inlinable
   public func multipliedFullWidth(by other: Self) -> (high: Self, low: Magnitude) {
-    // We define two utility functions for working with the high and low parts
-    // of the integer. Note that the low part is always unsigned, while the
-    // high part may be either signed or unsigned. Note also that we *represent*
-    // both in containers that are the full width of the original number, and
-    // let the compiler propagate the information about the ranges to which
-    // the values are clamped by the algorithm. There are likely some
-    // opportunities for improvement here, but there's no "half-width" type
-    // that we can access in a generic context to use here.
+    // We define a utility function for splitting an integer into high and low
+    // halves. Note that the low part is always unsigned, while the high part
+    // matches the signedness of the input type. Both result types are the
+    // full width of the original number; this may be surprising at first, but
+    // there are two reasons for it:
+    //
+    // - we're going to use these as inputs to a multiplication operation, and
+    //   &* is quite a bit less verbose than `multipliedFullWidth`, so it makes
+    //   the rest of the code in this function somewhat easier to read.
+    //
+    // - there's no "half width type" that we can get at from this generic
+    //   context, so there's not really another option anyway.
+    //
+    // Fortunately, the compiler is pretty good about propagating the necessary
+    // information to optimize away unnecessary arithmetic.
     func split<T: FixedWidthInteger>(_ x: T) -> (high: T, low: T.Magnitude) {
       let n = T.bitWidth/2
-      return (x &>> n, T.Magnitude(truncatingIfNeeded: x) & ((1 &<< n) &- 1))
-    }
-    func merge<T: FixedWidthInteger>(_ high: T, _ low: T.Magnitude) -> T {
-      let n = T.bitWidth/2
-      return high &<< n | T(truncatingIfNeeded: low)
+      return (x >> n, T.Magnitude(truncatingIfNeeded: x) & ((1 &<< n) &- 1))
```

Comment:
Why is the truncation init necessary here, if it was not below? Isn't `T.Magnitude` "wider" than `T` already? 

<sub>⚠️ I'm talking from perspective of new learner here, I really dunno what I'm talking about…</sub>

Reply:
It’s necessary here to handle negative values of ‘x’ correctly. That never happens in the other cases.

Reply:
OK, reading the comment on fa3a3d8, `init:truncatingIfNeeded` does, even though it's more verbose, actually produce less code by omitting the checks. You went with `init` below for better code clarity (less verbose), when you were certain the optimized will strip the checks. 

Swift's safety/performance and code brevity/clarity tradeoffs are quite counter intuitive here. But this does kill perf of unoptimized builds, right?

Reply:
Swift's taken the safety first mantra really seriously.

I mean, we use the more verbose API for less safe operations, but in the end, even when wielded by an expert like you, there is value in keeping the expressions mathematically concise, to give bugs less place to hide in cleaner code. 😢

It's as if the right way to code this was to express the idea first in most concise notation and *than* go in with an axe and butcher the clean math for performance's sake by adding all the unsafe `truncatungIfNeeded` garbage. But somehow it all feels wrong…

I'm always concerned with the performance impact on the unoptimized debug builds. I feel like we need some `@optimize(always)` attribute with some a compiler flag for really debugging this in development for such basic and performance critical building blocks as this. It always bothers me that `for-in` beats functional code with `map`/`reduce` just because they perform abysmally in unoptimized builds. 😤

Reply:
The perf for unoptimized builds can't be killed: it's already dead because inlining and hence specialization isn't going to happen for this function. That's OK though, it's still better than a fatalError (and short of adding a new @semantics annotation, we can't do anything about it yet).

Reply:
> It's as if the right way to code this was to express the idea first in most concise notation and then go in with an axe and butcher the clean math for performance's sake by adding all the unsafe truncatingIfNeeded garbage.

In the long-term, the right way to address this is probably to define some stdlib-internal integer types that always wrapping arithmetic. That's a useful building block to have for library development, and gets us to a best-of-both worlds situation. In the shorter term, keep in mind that `truncatingIfNeeded` and wrapping arithmetic are **not** just "turning off safety for performance reasons". They are semantically different operations, still with fully defined behavior; they are perfectly safe, and just mean something else.